xml_struct_derive/src/serialize/codegen.rs

```diff
@@ -0,0 +1,588 @@
+            as_attr_body: None,
+        }
+    }
+}
+
+/// The common sets of tokens which make up a `match` arm for an enum variant.
+struct VariantTokenSets {
+    /// The identifiers used for accessing the fields of an enum variant.
+    accessors: Vec<TokenStream>,
+
+    /// The calls for serializing the child nodes of the XML element
+    /// representing an enum variant.
+    content_calls: TokenStream,
+}
+
+/// Generates a list of accessors and set of calls to serialize content for an
+/// enum variant.
+fn generate_variant_token_sets(
+    name_tokens: TokenStream,
+    namespace_attrs: &[XmlAttribute],
+    fields: Vec<Field>,
+) -> VariantTokenSets {
+    let accessors: Vec<_> = fields
+        .iter()
+        .map(|field| &field.accessor)
+        .cloned()
+        .collect();
+
+    let Fields {
+        attr_fields,
+        child_fields,
+    } = partition_fields(fields);
+
+    let child_node_calls = generate_child_node_calls_for_fields(child_fields);
+
+    let content_calls = generate_xml_tag_calls(
+        name_tokens,
+        namespace_attrs,
+        &attr_fields,
+        Some(child_node_calls),
+    );
+
+    VariantTokenSets {
+        accessors,
+        content_calls,
+    }
+}
+
+/// Divides the fields of a struct or enum variant into those which will be
+/// represented as attributes and those which will be represented as child nodes.
+fn partition_fields(fields: Vec<Field>) -> Fields {
+    let (attr_fields, child_fields) = fields
+        .into_iter()
+        .partition(|field| matches!(field.props.repr, FieldRepr::Attribute));
+
+    Fields {
+        attr_fields,
+        child_fields,
+    }
+}
+
+/// Generates calls for adding namespace attributes to an element.
+fn generate_namespace_attrs_call(namespace_attrs: &[XmlAttribute]) -> TokenStream {
+    if !namespace_attrs.is_empty() {
+        let namespace_attrs: Vec<_> = namespace_attrs
+            .iter()
+            .map(|XmlAttribute { name, value }| quote!((#name, #value)))
+            .collect();
+
+        quote! {
+            .with_attributes([
+                #(#namespace_attrs,)*
+            ])
+        }
+    } else {
+        TokenStream::default()
+    }
+}
+
+/// Generates the calls necessary to serialize struct or enum fields as XML
+/// attributes.
+fn generate_attribute_field_calls(attr_fields: &[Field]) -> TokenStream {
+    if !attr_fields.is_empty() {
+        attr_fields
+            .iter()
+            .map(|field| {
+                let name = field_name_to_string_tokens(field);
+                let accessor = &field.accessor;
+                let ty = &field.ty;
+
+                quote! {
+                    <#ty as ::xml_struct::XmlSerializeAttr>::serialize_as_attribute(&#accessor, &mut start_tag, #name);
+                }
+            })
+            .collect()
+    } else {
+        TokenStream::default()
+    }
+}
+
+/// Generates calls to add a new XML element to the document, including any
```

Comment:
Vague in what way? "sets of tokens" is used elsewhere mostly because the functions in question generate multiple `TokenStream`s based on a single value. Here, I'm counting on the function signature to indicate "we're generating tokens" and the documentation to clarify what those tokens represent. I'm having trouble rewording this comment in a way that provides more information without being difficult to read.

Reply:
Actually ignore me, I don't think I was thinking this through enough